foundry.md

```diff
@@ -373,18 +373,67 @@ This rule takes the `address` value using `#asWord(#range(LM, ARGSTART +Int 4, 3
 
 ```k
     rule [call.getNonce]:
-          <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH RETSTART RETWIDTH => 1 ~> #push ~> #setLocalMem RETSTART RETWIDTH #bufStrict(32, NONCE) ... </k>
-          <output> _ => #bufStrict(32, NONCE) </output>
+          <k> CALL _ CHEAT_ADDR _VALUE ARGSTART _ARGWIDTH RETSTART RETWIDTH
+           => #loadAccount #asWord(#range(LM, ARGSTART +Int 4, 32))
+           ~> #foundryGetNonce ARGSTART RETSTART RETWIDTH
+           ~> #refund GCALL ...
+          </k>
+          <callGas> GCALL => 0 </callGas>
           <localMem> LM </localMem>
-          <account>
+       requires CHEAT_ADDR ==Int #address(FoundryCheat)
+        andBool #asWord(#range(LM, ARGSTART, 4)) ==Int 755185067 // selector ( "getNonce(address)" )
+     [priority(40)]
+
+
+     syntax KItem ::= "#foundryGetNonce" Int Int Int
+ // ----------------------------------------------------------
+    rule <k> #foundryGetNonce ARGSTART RETSTART RETWIDTH
```

Comment:
```suggestion
     syntax KItem ::= "#foundryGetNonce" Int Int Int
 // ------------------------------------------------
    rule <k> #foundryGetNonce ARGSTART RETSTART RETWIDTH
```

Reply:
I think we can just call this `#getNonce` instead of `#foundryGetNonce`.

Also, we should compute the `ACCTID` _before_ ppassing it into this function, because currently we're doing a map-lookup on the accounts cell (with key `<ACCTID>`), for a value that is set in the side-conditino. This is very computationally exensive, because the backend will have to search through every element in teh `<accounts>` cell looking for one with a key that matches.

Instead, the previous rule for `CALL` should have:

```
~> #getNonce #asWord(#range(LM, ARGSTART +Int 4, 32)) RETSTART RETWIDTH
```

So that this one can just become:

```
#getNonce ACCTID RETSTART RETWIDTH => #setLocalMem RETSTART RETWIDTH #bufStrict(32, NONCE)
```

without the side-condition `ACCTID ==Int ...` (it's precomputed in the previous rule). This is much more efficient, we can do direct lookup on the key instead of a search.

Reply:
Makes sense! I was just trying to follow the structure of the rules for `load` and `store`. Maybe it makes sense to also optimize those rules. I can do that in this PR, or alternatively, create a new one.

Reply:
Issue created: https://github.com/runtimeverification/evm-semantics/issues/1397